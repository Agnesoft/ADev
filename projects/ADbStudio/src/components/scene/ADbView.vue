<template>
    <div class="adb-view" :class="{'search-mode': searchedWord !== ''}">
        <div class="view">
            <Edge v-for="(edges,key) in edgesGrouped" :key="key" :elements="edges" :viewSize="viewSize" :class="{selected: edgeIsSelected(edges)}" v-on="$listeners" />
            <Node v-for="node in nodes" :key="node.id" :element="node" :viewSize="viewSize" :class="{selected: node.id == searchedWord}" v-on="$listeners" />
        </div>
    </div>
</template>

<script>
import Node from "./Node";
import Edge from "./Edge";
import { mapGetters } from 'vuex'

export default {
    name: "ADbView",
    components: {
        Node,
        Edge,
    },
    data(){
        return {
            viewWidth: 0,
            viewHeight: 0,
        }
    },
    props: {
        searchedWord: String,
    },
    computed: {
        ...mapGetters({
            nodes: 'scene/getNodes',
            edges: 'scene/getEdges',
            edgesGrouped: 'scene/getEdgesGrouped',
        }),
        viewSize(){
            return {
                width: this.viewWidth,
                height: this.viewHeight
            }
        }
    },
    methods: {
        onResize() {
            this.viewWidth = document.documentElement.clientWidth;
            this.viewHeight = document.documentElement.clientHeight;
        },
        edgeIsSelected(edges){
            return edges.find((edge) => edge.id == this.searchedWord) ? true : false;
        }
    },
    mounted() {
        window.addEventListener('resize', this.onResize)
        this.onResize();
    },

    beforeDestroy() {
        window.removeEventListener('onmousewheel', this.onResize)
    }

}
</script>

<style scoped>
    .adb-view{
        position: relative;
        transform: scale(var(--scale,1)) translate(var(--move-x,0), var(--move-y,0));
        width: 100%;
        height: 100%;
    }
</style>
