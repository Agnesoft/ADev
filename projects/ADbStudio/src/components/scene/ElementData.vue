<template>
    <FadeTransition :opened="show">  
        <table v-if="dataValid" class="element-data" :style="dataStyle">
            <tr v-for="(value, data_key) in data" :key="data_key">
                <td>{{ data_key }}</td>
                <td>{{ value }}</td>
            </tr>
        </table>
        <div v-else class="element-data" :style="dataStyle">
            {{ $t("no-data-selected") }}
        </div>
    </FadeTransition>
</template>

<script>
import FadeTransition from "@/components/transitions/FadeTransition";

export default {
    name: "ElementData",
    components: {
        FadeTransition,
    },
    props: {
        show: {
            type: Boolean,
            default: true,
        },
        data: Object,
        dataStyle: {
            type: Object,
            default() {return {}},
        }
    },
    computed: {
        dataValid(){
            if(Object.keys(this.data).length != 0){
                return true;
            }
            return false;
        }
    }
}
</script>

<style scoped>
    .element-data{
        background: var(--light-color);
        box-shadow:  0 0 3px var(--primary-color), 2px 2px 5px var(--dark-color);
        /* border-radius: 0.4rem; */
        border: solid 1px var(--primary-color);
        padding: 0.4rem;
        width: 12rem;
        position: relative;
        z-index: 1000;
    }
    .element-data td,.element-data th {
        border: 1px solid #dddddd;
        text-align: left;
        padding: 0.4rem 0.8rem;
    }
    .element-data tr:nth-child(odd) {
        background-color: var(--light-color);
    }
    .element-data tr:nth-child(even) {
        background-color: #dddddd;
    }
    .element-data tr:nth-child(even) td{
        border: 1px solid var(--light-color);
    }
    table.element-data {
        border-collapse: collapse;
        border-style: hidden;
    }

</style>
