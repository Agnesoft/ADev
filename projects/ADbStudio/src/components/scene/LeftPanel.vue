<template>

    <SlideXTransition :opened="has_data"  >
        <div class="left-panel">
            <div class="tabs">
                <div class="tab-button" :class="{active: tab.key==active_tab_key}" v-for="tab in tabs" :key="tab.key" :title="tab.title" >
                    <BaseButton class="btn-circle" :class="{'btn-primary' : tab.key==active_tab_key,'btn-light' : tab.key!=active_tab_key}" @click="buttonClicked(tab.key)">
                        <BaseIcon :icon_key="tab.key" />
                    </BaseButton>
                </div>
            </div>
            <SlideXTransition :opened="active_tab_key != -1"  >
                <div class="tab-box">
                    <header class="tab-header">
                        {{tab_title}}
                    </header>
                    <transition name="fade" mode="out-in">
                        <component class="tab-content" v-bind:is="component_content"  v-on="$listeners"></component>
                    </transition>
                </div>
            </SlideXTransition>    
        </div>
    </SlideXTransition> 
</template>

<script>
import SlideXTransition from "../transitions/SlideXTransition";
import { mapGetters } from 'vuex'

export default {
    name: "LeftPanel",
    components: {
        SlideXTransition,
        NodeListData: () => import('./NodeListData'),
        EdgeListData: () => import('./EdgeListData'),
        RawData: () => import('./RawData')
    },
    data() {
        return {
            active_tab_key: -1,
            tabs: [
                {
                    key: 'nodes',
                    title: this.$t('list-of-nodes'),
                    component_content: "NodeListData",
                },
                {
                    key: 'edges',
                    title: this.$t('list-of-edges'),
                    component_content: "EdgeListData",
                },
                {
                    key: 'data',
                    title: this.$t('raw-data'),
                    component_content: "RawData",
                },
            ]
        }
    },
    computed: {
        ...mapGetters({
            data: 'scene/getData',
        }),
        active_tab(){
            if(!this.active_tab_key){
                return false;
            }
            let active_tab_key = this.active_tab_key;
            let tab = this.tabs.find((elem) => elem.key == active_tab_key);
            if(!tab){
                return false;
            }
            return tab;
        },
        component_content(){
            let active_tab_key = this.active_tab_key;
            if(active_tab_key){
                return this.active_tab.component_content;
            }
            return false;
        },
        tab_title(){
            let active_tab_key = this.active_tab_key;
            if(active_tab_key){
                return this.active_tab.title;
            }
            return '';
        },
        has_data(){
            return !(Object.keys(this.data).length === 0 && this.data.constructor === Object);
        }
    },
    methods: {
        buttonClicked(tab_key){
            let tab = this.tabs.find((elem) => elem.key == tab_key);
            if(!tab){
                return;
            }
            if(this.active_tab_key === tab_key){
                this.active_tab_key = -1;
            } else {
                this.active_tab_key = tab_key;
            }
        },
    }
}
</script>

<style scoped>
    .left-panel{
        display: flex;
    }
    .tabs{
        display: flex;
        flex-direction: column;
        margin-right: var(--grid-gap);
    }
    .btn{
        margin: 0.2rem;
        transition: all 0.2s ease;
    }
    .tab-box{
        border-radius: 0.4rem;
        overflow: hidden;
        background: white;
        display: flex;
        flex-direction: column;
        max-height: 80vh;
        --shadow: var(--dark-color) 1px 1px 5px 0px;
        -webkit-box-shadow:  var(--shadow);
        -moz-box-shadow:  var(--shadow);
        box-shadow:  var(--shadow);
    }
    .tab-header{
        padding: 0.5rem;
        white-space: nowrap;
    }
    .tab-content{
        width: 15rem;
        padding: 0.5rem;
        margin: 0;
        height: 100%;
        overflow: auto;
    }
    .fade-enter-active {
        transition: all 0.3s
    }
    .fade-enter {
        opacity: 0;
    }
    .fade-leave-active {
        transition: all 0.3s;
        opacity: 0;
    }
    .tab-header{
        background: var(--primary-color);
        color: var(--light-color);
    }
</style>
