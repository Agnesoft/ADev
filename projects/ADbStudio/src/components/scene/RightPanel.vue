<template>
    <SlideXTransition :opened="elements.length > 0"  >
        <div class="right-panel">
            <div>
                <BaseButton class="btn-primary" @click="removeAllElements">
                    {{ $t('hide-all-data') }}
                </BaseButton>
                
                <transition-group name="fade" >
                    <BaseBox  v-for="element in elements" :key="element.id" :def_opened="true" :closable="true">
                        <h4 slot="title">{{ getTitle(element) }}</h4>
                        <div class="controls" slot="controls">
                            <BaseButton class="btn-light" @click="closeBox(element.id)">
                                <svg width="100%" height="100%">
                                    <use xlink:href="../../assets/images/icons.svg#close" x = 0 y = 0></use>
                                </svg>
                            </BaseButton>
                        </div>
                        <ElementData slot="content" v-if="dataValid(element)" :data="element.data" />
                        <div slot="content" v-else class="no-data">
                            {{ $t("no-data-selected") }}
                        </div>
                    </BaseBox>
                </transition-group>
            </div>
        </div>
    </SlideXTransition>     
</template>

<script>
import SlideXTransition from "../transitions/SlideXTransition";
import ElementData from './ElementData';
import { mapActions,mapGetters } from 'vuex';

export default {
    name: "RightPanel",
    components: {
        SlideXTransition,
        ElementData,
    },
    data(){
        return {
        }
    },
    computed: {
        ...mapGetters({
            elements: 'scene/getElementsData',
        }),
    },
    methods: {
        ...mapActions({
            removeElement: 'scene/removeElementData',
            removeAllElements: 'scene/removeAllElementsData',
        }),
        dataValid(element){
            if(Object.prototype.hasOwnProperty.call(element, 'data') && Object.keys(element.data).length != 0){
                return true;
            }
            return false;
        },
        getTitle(element){
            if(element.id < 0){
                return this.$t('edge-id',{'id': element.id});
            } else {
                return this.$t('node-id',{'id': element.id});
            }
        },
        closeBox(id){
            this.removeElement({id: id});
        }
    }
}
</script>

<style scoped>
    .fade-enter-active {
        transition: all 0.3s
    }
    .fade-enter {
        opacity: 0;
    }
    .fade-leave-active {
        transition: all 0.3s;
        opacity: 0;
    }
    .right-panel{
        max-width: 0.3vw;
        min-width: 12rem;
        max-height: 100%;
        position: relative;
        overflow: auto;
    }
    .right-panel>div{
        background: white;
        position: absolute;
        top: 0;
        right: 0;
        width: 100%;
    }

</style>
