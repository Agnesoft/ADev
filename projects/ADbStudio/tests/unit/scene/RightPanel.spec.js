import { mount  } from '@vue/test-utils';
import Vuex from 'vuex';
import RightPanel from "@/components/scene/RightPanel";
import ADbView from "@/components/scene/ADbView";
import LeftPanel from "@/components/scene/LeftPanel";
import ElementData from "@/components/scene/ElementData";
import scene_store from '../../../src/store/modules/scene'
//import NodeInfo from "@/components/scene/NodeInfo";

import { localVue, i18n,scene_state } from '../TestUtils'


describe('Scene',() => {
    let wrapper;
    let store;
    let slideTransitionStub; 
    let transtionStub;

    beforeEach(() => {
        slideTransitionStub = () => ({
            props: {
                opened: Boolean,
            },
            template: '<div v-if="opened" ><slot></slot></div>',
          });
          transtionStub = () => ({
              template: '<div><slot></slot></div>',
            });
        
        localVue.use(Vuex)
        store = new Vuex.Store({
            modules: {
                scene: {
                    namespaced: true,
                    actions: scene_store.actions,
                    mutations: scene_store.mutations,
                    state: scene_state,
                    getters: scene_store.getters
                }
              }
        });
        
        wrapper = mount(RightPanel, {
            localVue,
            i18n,
            store,
            stubs: {
                SlideXTransition:  slideTransitionStub(),
                transition:  transtionStub(),
                "transition-group":  transtionStub(),
            }
        });
    })
    it('renders all elements', () => {
        expect(wrapper.find(".btn-primary").exists()).toBe(true);
        expect(wrapper.find(".btn-primary").text()).toBe("Hide all data");
        expect(wrapper.findAll(".base-box").length).toBe(2);
    })
    it('renders correctly (snapshot)', () => {
        expect(wrapper.element).toMatchSnapshot()
    })
    it('renders boxes correctly',() => {        
        let boxes = wrapper.findAll(".base-box");
        expect(boxes.at(0).find('.title').text()).toBe('Node 1');
        let firsttable = boxes.at(0).find('.content').text();
        expect(firsttable).toBe('name Jananumber 18');

        expect(boxes.at(1).find('.title').text()).toBe('Edge -2');
        let secondtable = boxes.at(1).find('.content').text();
        expect(secondtable).toBe('foo bar');
    })
    it('closes box after clicking close box button', async () => {
        let boxes = wrapper.findAll(".base-box");
        await boxes.at(0).find('.controls>.btn').trigger('click');
        expect(wrapper.findAll(".base-box").length).toBe(1);
    })
    it('closes all boxes after clicking close all button', async () => {
        await wrapper.find(".btn-primary").trigger('click');
        expect(wrapper.findAll(".base-box").length).toBe(0);
    })

   
})
