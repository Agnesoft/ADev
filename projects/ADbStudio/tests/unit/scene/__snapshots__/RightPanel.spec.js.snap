// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scene renders correctly (snapshot) 1`] = `
<div>
  <div
    class="right-panel"
  >
    <div>
      <button
        class="btn btn-primary"
      >
        
                Hide all data
            
      </button>
       
      <div
        name="fade"
      >
        <div
          class="base-box"
        >
          <div
            class="header"
          >
            <div
              class="title"
            >
              <h4>
                Node 1
              </h4>
            </div>
             
            <button
              class="btn btn-light toggle-button opened"
            >
              <svg
                class="base-icon"
                height="100%"
                version="1.1"
                width="100%"
                xmlns="http://www.w3.org/2000/svg"
                xmlns:xlink="http://www.w3.org/1999/xlink"
              >
                <use
                  x="0"
                  xlink:href="../../assets/images/icons.svg#arrow"
                  y="0"
                />
              </svg>
            </button>
             
            <div
              class="controls"
            >
              <button
                class="btn btn-light"
              >
                <svg
                  height="100%"
                  width="100%"
                >
                  <use
                    x="0"
                    xlink:href="../../assets/images/icons.svg#close"
                    y="0"
                  />
                </svg>
              </button>
            </div>
          </div>
           
          <div
            name="slide-y"
          >
            <div
              class="content"
            >
              <table
                class="element-data"
              >
                <tr>
                  <td>
                    name
                  </td>
                   
                  <td>
                    Jana
                  </td>
                </tr>
                <tr>
                  <td>
                    number
                  </td>
                   
                  <td>
                    18
                  </td>
                </tr>
              </table>
            </div>
          </div>
        </div>
        <div
          class="base-box"
        >
          <div
            class="header"
          >
            <div
              class="title"
            >
              <h4>
                Edge -2
              </h4>
            </div>
             
            <button
              class="btn btn-light toggle-button opened"
            >
              <svg
                class="base-icon"
                height="100%"
                version="1.1"
                width="100%"
                xmlns="http://www.w3.org/2000/svg"
                xmlns:xlink="http://www.w3.org/1999/xlink"
              >
                <use
                  x="0"
                  xlink:href="../../assets/images/icons.svg#arrow"
                  y="0"
                />
              </svg>
            </button>
             
            <div
              class="controls"
            >
              <button
                class="btn btn-light"
              >
                <svg
                  height="100%"
                  width="100%"
                >
                  <use
                    x="0"
                    xlink:href="../../assets/images/icons.svg#close"
                    y="0"
                  />
                </svg>
              </button>
            </div>
          </div>
           
          <div
            name="slide-y"
          >
            <div
              class="content"
            >
              <table
                class="element-data"
              >
                <tr>
                  <td>
                    foo
                  </td>
                   
                  <td>
                    bar
                  </td>
                </tr>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
